Updated MongoDB tutorial

VatsalMehta27 opened the pull request:
Refined and updated some of the original examples/explanations, and added a section to discuss `populate`.

VatsalMehta27 committed:
updated mongodb tutorial examples and added populate explanation

mwand commented:
"All documents in this collection must follow the schema defined by
kittySchema." 

maybe add: Notice that Kitten is a constructor (like a class name), so
we write it in upper case. 

==================================

const q = await QuestionModel.findOneAndUpdate(
  { _id: new ObjectId(qid) },
  { $addToSet: { views: username } },
  { new: true }
);



I don't understand what the new ObjectId(...) is doing here.  It looks
like this code is trying to find a document with a brand-new ObjectId
(which presumably doesn't exist yet).

What is the type of qid?  Why does it say new ObjectId(qid) rather than
qid?  

The alternative version that follows has the same problem.


==========================================

const posts = await Post.find().populate("author");

If there were more than one Post in the collection, would this
populate all of them?

===========================================

Each Post references a User in the author field by the ObjectID. The schemas and models would be defined as:

const postSchema = new mongoose.Schema({
  title: String,
  content: String,
  author: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
});

==
Probably worth saying that you can't write 

const postSchema = new mongoose.Schema({
  title: String,
  content: String,
  author: User
});

because User (or anything like it) is not a legal type to appear in a
Schema definition.

===============================================

placese => places

=========================

The tutorial example (zip file) needs some loving-care.   How are the
four files related?

books.ts:

line 7 appears to have some lost text at the end.

Why is there a bunch of stuff at the end that is commented out?  (It's
probably stuff that I left in there, but you should edit it).

line 178: authors appears to be undefined.

example.ts:

remove commented lines 58-59.

populate.ts:

seems to duplicate example.ts.  Is it a cleaned-up
version of example.ts?  I wrote these files, but I don't remember,
sorry.

line 79: what's that .exec() mean?  Is there a way of doing without
it?

stories.ts:

More duplication.

Can you edit these files into a single coherent one?


console.log("done, hit Ctrl-C to exit")
Please explain why this is needed here, but not in the codebase.

VatsalMehta27 committed:
Updated tutorial with feedback

VatsalMehta27 committed:
updated tutorial code

VatsalMehta27 commented:
> "All documents in this collection must follow the schema defined by kittySchema."
> 
> maybe add: Notice that Kitten is a constructor (like a class name), so we write it in upper case.

Updated.

 ==================================

> const q = await QuestionModel.findOneAndUpdate( { _id: new ObjectId(qid) }, { $addToSet: { views: username } }, { new: true } );
> 
> I don't understand what the new ObjectId(...) is doing here. It looks like this code is trying to find a document with a brand-new ObjectId (which presumably doesn't exist yet).
> 
> What is the type of qid? Why does it say new ObjectId(qid) rather than qid?
> 
> The alternative version that follows has the same problem.

`qid` has type string, and the new ObjectID converts that string to an ObjectID instance. However, it's not necessary because Mongoose does it internally, and I've removed it from the examples.

==========================================

> const posts = await Post.find().populate("author");
> 
> If there were more than one Post in the collection, would this populate all of them?

Yes, I've added a sentence to clarify this.

 ===========================================

> Each Post references a User in the author field by the ObjectID. The schemas and models would be defined as:
> 
> const postSchema = new mongoose.Schema({ title: String, content: String, author: { type: mongoose.Schema.Types.ObjectId, ref: "User" }, });
> 
> == Probably worth saying that you can't write
> 
> const postSchema = new mongoose.Schema({ title: String, content: String, author: User });
> 
> because User (or anything like it) is not a legal type to appear in a Schema definition.

Added as in-line comment within example.

===============================================

> placese => places

Updated.

 =========================
 
> The tutorial example (zip file) needs some loving-care. How are the four files related?
> 
> books.ts:
> 
> line 7 appears to have some lost text at the end.
> 
> Why is there a bunch of stuff at the end that is commented out? (It's probably stuff that I left in there, but you should edit it).
> 
> line 178: authors appears to be undefined.
> 
> example.ts:
> 
> remove commented lines 58-59.
> 
> populate.ts:
> 
> seems to duplicate example.ts. Is it a cleaned-up version of example.ts? I wrote these files, but I don't remember, sorry.
> 
> line 79: what's that .exec() mean? Is there a way of doing without it?
> 
> stories.ts:
> 
> More duplication.
> 
> Can you edit these files into a single coherent one?
> 
> console.log("done, hit Ctrl-C to exit") Please explain why this is needed here, but not in the codebase.

I did an overhaul of the tutorial, which is uploaded as a zip under assets. I changed the example to use Kitty, Vet, and Clinic schemas, to showcase both find and populate queries. I hadn't previously looked at it.

I also added other setup-related stuff, like a package.json and linting, so that students can run it easily. There are two ways to run the tutorial: `quick-tutorial` which runs all the queries and outputs, and `step-by-step` which executes one query at a time and waits for the student to read before continuing. I added the second option because there were too many examples to go through when executed together, in my opinion. This is all explained in the README file.

What are your thoughts on the project folder structure? I found having all the code in one file, like the previous examples, was too long, so I separated it out into different files. I can refactor based on what you see fit.

mwand commented:
Tutorial code looks great!  Just a few comments:

1. I got  an ECONNREFUSED the first time I ran it, because my MongoDB service wasn't running.   You might put something in the readme about this, or  put the mongoose.connect in a try/catch to print out a useful error message.
2. One thing I covered in my tutorial was how sometimes TS will fail to reject a query with a type error, because ObjectID is untyped. (Not sure if that's worth fixing in this version).
3. The script would be cooler if for each example it printed out the query along with the result.  Also not sure if that's worth adding right now.

Anyhow, great job!

VatsalMehta27 committed:
updated tutorial with feedback

VatsalMehta27 committed:
added spacing for easier reading
